features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,248 @@
-# next webhooks design
+# Background
 
-## Overview
+Web-hooks would be used to notify external services of events in auth gear.
 
-skygear sends webhooks for events that happen in your app, and are designed to help a developer to monitor events, sync database, and many other use cases.
 
-When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+# Proposed Designs
 
-## SYNC and ASYNC hooks
+## Events
+Web-hook events are triggered when some mutating operation is performed by
+auth gear.
 
-There are two types of webhook:
+Each operation will trigger two events: BEFORE and AFTER:
+- BEFORE events would be triggered before the operation is performed; the
+  operation can be failed by web-hook handler.
+- AFTER events would be trigger after the operation is performed.
 
-- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
-- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+Both events has the same event payload.
 
-## Request Headers
 
-Following headers will be added to webhook POST request:
+## Delivery
 
-- `X-Skygear-Webhook-Req-ID`: an ID of the request.
-- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
-- `X-Skygear-Webhook-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+The events would be delivered to web-hook handlers as an HTTP POST request to
+the handler endpoint.
 
-## Request Payload
+The handler endpoint can be specified as full URL or an absolute path. Absolute
+paths will be resolved to full URL using an inferred URL scheme and authority:
+- from HTTP request URL if the event is generated from it; otherwise,
+- from the tenant configuration.
 
-Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+Each event type can have multiple handlers; the order of deliveries is
+unspecified.
 
-All event payloads follows following payload format:
+BEFORE events will always be delivered before AFTER events. BEFORE events will
+be delivered in well-defined order during a request. AFTER events will be
+delivered in unspecified order.
 
-```json=
+Web-hook handler must be idempotent.
+
+Web-hook handler must return a status code within the 2XX range. Responding with
+status code outside the range, including 3XX & 5XX, would be considered a
+failed delivery.
+
+The time spent in a delivery must not exceed 5 seconds, otherwise would be
```

Comment:
I'll say 5 seconds is already too long; most interactive production web-site would abort the request much faster (e.g. in 1 sec) to prevent long-lasting request from hogging the resources. For example, GitHub's unicorn page.

Reply:
Can we make it configurable? I agree in 70% of case, 1 sec is too long, but there are also occasion people might want longer timeout. 

Another factor is for async webhook, the timeout could be even longer.

Reply:
If we allow setting a longer timeout, people will almost always set it to be the longest possible. So I prefer it to be non-configurable.

If the handler takes more than 5 seconds, it is more than likely some nontrivial work is being performed (e.g. intensive computation, heavy database query). User should perform these work in background instead.

Reply:
I know it is "better" to perform some long task in the background instead.

However, it is a decision we need to make for the pros and cons trade-off (pros and cons of having a short fixed default timeout):

* For before sync hooks:
    * Pros:
        * Avoid disastor since it could be locking some resources
        * Improve end-users' latency
    * Cons:
        * Developers need to write the code carefully
        * Developers may encounter unexpected error for short timeout, worse experience
* For after async hooks:
    * Pros:
        * Improve end-users' latency
    * Cons:
        * Need developer awareness to write background job code

Let me know if you think the above list missed anything. Otherwise I think it is not so obvious for after async hooks that a short fixed timeout is better. We can left the trade-off to users (hence make the timeout configurable)

@kiootic I hope the above help you make more design decision in future and improve our quality of discussion here.

----

Or another way to think about it is, you also said "people will almost always set it to be the longest possible", that meant there is a strong reason why people want to do it, and we need to take that into account. Why people almost always set it??? Is that reason really bad enough that we should dismissed it?

The subjective decision here is if the risk of "locking resources and affect scalability of the system" meant we should take away the choice from users. In this case I think it is, I have to admit it is gut feeling but if I need to provide a reason, one could be because the cluster is multi-tenant, it could take down other people's app at the same time.

Reply:
For after async events: they're delivered asynchronously, so the request/response latency is not affected.

I'm considering the worst case here: people will likely take advantage of given resources. Also, let's say the default timeout is 1 sec and maximum 5 sec: if developer encountered the timeout, they'll almost surely just increase the timeout, instead of considering whether they're doing too much work in the handler.

The multi-tenant nature of auth gear is indeed my main reason that I prefer it to be a short and non-configurable timeout.

Reply:
> For after async events: they're delivered asynchronously, so the request/response latency is not affected.

Oh I should have separate it, the latency is not just requested latency, it could be a developer well aware of the timeout and write better code and hence their eventually consistency whatever counter respond faster. So I included that on my comparison table. But yea I should have specify "request latency" and "latency" separately.



> I'm considering the worst case here: people will likely take advantage of given resources. Also, let's say the default timeout is 1 sec and maximum 5 sec: if developer encountered the timeout, they'll almost surely just increase the timeout, instead of considering whether they're doing too much work in the handler.

Yea but again, if we consider the pros/cons, in the case of after sync hook, the only resources we got "locked" is the connection if the users keep increase the timeout to maximum. That won't bother us much?

Reply:
> Yea but again, if we consider the pros/cons, in the case of after sync hook, the only resources we got "locked" is the connection if the users keep increase the timeout to maximum. That won't bother us much?

I think you mean after async? I'm not sure if the connection socket is the only resource user's holding while the event delivery is in progress: we need to consider memory, processing power, bandwidth, storage, etc.

Even if we concluded that the amount of resource is trivial, do we want to encourage developer to do incorrect thing (long-running job in event handler), by providing an easy way out (increasing timeout)?

Reply:
So it sounds like my assumption of after async hook would only involve connection resources on our side is right? In that case my recommendation would be:

1. For before sync webhook, we have a default value of 5 sec and it is not configurable (@louischan-oursky suggested 5 secs is too short originally, so he may have another opinion on this)
2. For after async webhook, we have a default value of 30 sec and it is configurable to at most 60 secs.

> Even if we concluded that the amount of resource is trivial, do we want to encourage developer to do incorrect thing (long-running job in event handler), by providing an easy way out (increasing timeout)?

As explained above, we need to consider the pros and cons of each design decision, not just gut feeling of rights or wrong. In this case, the "easy way out" is a major improvement to developer experience when they needed. The "incorrect thing" do not have any major drawback besides the resources to hold the connection -- Unless there are more reason to justify why it is an "incorrect thing" for async hook to be longer

Reply:
I'd agree on a longer timeout for after async event deliveries, even 60 seconds would be okay by default; I mainly oppose configurable/long timeout for before sync events.

Discussed with @louischan-oursky :
- For before sync events: 5 seconds by default, configurable in app config (can be modify by admin only)
- For after async events: 60 seconds, non-configurable for now.

Reply:
Agree if before sync is configurable in app config by admin only, seems more flexible especially in the case if the Skygear Cluster is owned by, say an Enterprise, they can do whatever...

But I'm curious, the main reason for after async is non-configurable per app is....? Is there any reason why one might want to shorten it?

Reply:
60 seconds is more than enough for most use cases, so just make it non-configurable for now.

Reply:
Sounds fine.